bpo-6721: Sanitize logging locks while forking

gpshead opened the pull request:
This part is simple, acquire and release the logging lock and each logging handler's lock around fork.

A unittest is still needed (relatively trivial to create).

The `logging.handlers` module has a `QueueHandler` and `QueueListener` which uses the `queue` module... that is full of locks.  I'm not addressing that as I don't see an immediately obvious correct thing to do with that queue.  A QueueListener thread won't be running in the forked child anyways so the right thing for anyone using one of those is likely to remove the QueueHandler from their logging config entirely... unless they go on to re-establish their complicated queue based logging handler setup in the child.  suggestion: Do nothing.  Let people using QueueHandler deal with their own problems as their application sees fit.

<!-- issue-number: bpo-6721 -->
https://bugs.python.org/issue6721
<!-- /issue-number -->

gpshead committed:
[bpo-6721] Sanitize logging locks while forking.

pitrou commented on Lib/logging/__init__.py:
```diff
@@ -793,6 +798,9 @@ def createLock(self):
         Acquire a thread lock for serializing access to the underlying I/O.
         """
         self.lock = threading.RLock()
+        os.register_at_fork(before=self.acquire,
```
This may register an arbitrary number of callbacks across an application's lifetime, also it makes the handles eternal by keeping a reference to all of them. Instead, we could have a global WeakSet of handlers and a single set of callbacks that iterate over the live handlers.

gpshead replied:
Another option would be a `before=lambda acquire=weakref.WeakMethod(self.acquire): acquire()()`... I'm used to applications that don't go creating tons of `logging.Handler` instances and throwing them away all over the place. :)

I think I'll go with the WeakSet approach to avoid an every growing list of callables in that strange Handler discarding application case...

gpshead committed:
Use the WeakSet approach per code review.

gpshead committed:
blurb

gpshead committed:
Make the logic conditional on register_at_fork.

gpshead committed:
Add a unittest!

gpshead committed:
Actually use the registration fn...

gpshead committed:
Fix the test to actually work.

Now it fails on the previous code and is fixed by the logging changes.
The test now uses a thread to hold the locks during the fork in a
synchronized manner.

Due to mixing of fork and a thread there is potential for deadlock
in the child process.  buildbots and time will tell if this actually
manifests itself in this test or not.  :/
